url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Node.java

```diff
@@ -0,0 +1,133 @@
+    	
+// 	   SearchResults results = node13.search(["boyzone"]); // this message would route to boyzone node
+        Vector<SearchResponse> responses = controller.sendSearch(words);
+        if(responses.isEmpty())return (new SearchResult[0]);
+        // A big constraint: a node cannot be waiting for more than one search. Otherwise it is hard to categorise
+        // the search responses. It can still receive index requests and the others while its waiting though.
+        ArrayList<SearchResult> results = new ArrayList<SearchResult>();
+        //reassemble responses to results
+        ArrayList<ResultPart> intermediaryResults = new ArrayList<Node.ResultPart>();
+        for(SearchResponse response: responses){
+        	for(Url url:response.urls){
+        		intermediaryResults.add(new ResultPart(response.word, url.url, url.rank));//makes my life easier converting
+        	}
+        }
+        Collections.sort(intermediaryResults);
+        String curr = intermediaryResults.get(0).url;
+
+        //the following loop partitions the urls. It is similar to a functional language takeWhile
+        // in this case it builds up a SearchResult as long as the next url is the same as the current.
+        //its a bit of hack but it is hard to do comparisons on static arrays such as the words[] in SearchResult
+        //while it is building. I cannot check search_results.get(current_url).words.contains(word)
+        //which would have made my life easier
+        ArrayList<String> urlWords = new ArrayList<String>();
+        int urlFrequency = 0;
+        SearchResult newResult = new SearchResult();
+        for(int i = 0; i < intermediaryResults.size(); i++) {
+        	ResultPart next = intermediaryResults.get(i);
+			if(next.url.equals(curr)){
+				urlFrequency += next.rank;
+				urlWords.add(next.word);
+			}else{				
+				newResult.url = curr;
+				newResult.words = (String[])urlWords.toArray();
+				newResult.frequency = urlFrequency;
+				results.add(newResult);
+				
+				curr = next.url;
+				urlFrequency = next.rank;
+				urlWords.clear();
+				urlWords.add(next.word);
+				newResult = new SearchResult();
+			}
+		}
+        //the cleanup after the last iteration of the loop
+        newResult.words = (String[])urlWords.toArray();
+        newResult.frequency = urlFrequency;
+        results.add(newResult);
+        
+        return (SearchResult[])results.toArray();
+      //Stick to the interface, brother. Stick to the interface.
+    }
+    private class ResultPart implements Comparable<ResultPart>{
+    	String word,url;
+    	int rank;
+    	ResultPart(String w, String u, int r){
+    		word = w; url = u; rank = r;
+    	}
+		//we want to sort them based on url so that the same urls in an array align next to each other
+		public int compareTo(ResultPart obj) {
+			return url.compareTo(obj.url);
+		}
+		@Override
+		public boolean equals(Object obj) {
+			if(obj instanceof ResultPart){
+				return (url == ((ResultPart)obj).url);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.